server/msgtrace.go

```diff
@@ -0,0 +1,677 @@
+	v, ok := e.(*T)
+	if ok {
+		return v
+	}
+	return nil
+}
+
+func (t *MsgTraceEvent) Ingress() *MsgTraceIngress {
+	if len(t.Events) < 1 {
+		return nil
+	}
+	return getTraceAs[MsgTraceIngress](t.Events[0])
+}
+
+func (t *MsgTraceEvent) SubjectMapping() *MsgTraceSubjectMapping {
+	for _, e := range t.Events {
+		if e.typ() == MsgTraceSubjectMappingType {
+			return getTraceAs[MsgTraceSubjectMapping](e)
+		}
+	}
+	return nil
+}
+
+func (t *MsgTraceEvent) StreamExports() []*MsgTraceStreamExport {
+	var se []*MsgTraceStreamExport
+	for _, e := range t.Events {
+		if e.typ() == MsgTraceStreamExportType {
+			se = append(se, getTraceAs[MsgTraceStreamExport](e))
+		}
+	}
+	return se
+}
+
+func (t *MsgTraceEvent) ServiceImports() []*MsgTraceServiceImport {
+	var si []*MsgTraceServiceImport
+	for _, e := range t.Events {
+		if e.typ() == MsgTraceServiceImportType {
+			si = append(si, getTraceAs[MsgTraceServiceImport](e))
+		}
+	}
+	return si
+}
+
+func (t *MsgTraceEvent) JetStream() *MsgTraceJetStream {
+	for _, e := range t.Events {
+		if e.typ() == MsgTraceJetStreamType {
+			return getTraceAs[MsgTraceJetStream](e)
+		}
+	}
+	return nil
+}
+
+func (t *MsgTraceEvent) Egresses() []*MsgTraceEgress {
+	var eg []*MsgTraceEgress
+	for _, e := range t.Events {
+		if e.typ() == MsgTraceEgressType {
+			eg = append(eg, getTraceAs[MsgTraceEgress](e))
+		}
+	}
+	return eg
+}
+
+const (
+	errMsgTraceOnlyNoSupport   = "Not delivered because remote does not support message tracing"
+	errMsgTraceNoSupport       = "Message delivered but remote does not support message tracing so no trace event generated from there"
+	errMsgTraceNoEcho          = "Not delivered because of no echo"
+	errMsgTracePubViolation    = "Not delivered because publish denied for this subject"
+	errMsgTraceSubDeny         = "Not delivered because subscription denies this subject"
+	errMsgTraceSubClosed       = "Not delivered because subscription is closed"
+	errMsgTraceClientClosed    = "Not delivered because client is closed"
+	errMsgTraceAutoSubExceeded = "Not delivered because auto-unsubscribe exceeded"
+)
+
+type msgTrace struct {
+	ready int32
+	srv   *Server
+	acc   *Account
+	// Origin account name, set only if acc is nil when acc lookup failed.
+	oan   string
+	dest  string
+	event *MsgTraceEvent
+	js    *MsgTraceJetStream
+	hop   string
+	nhop  string
+	dm    bool // Deliver message to client/LN/service import/etc...
+	ct    compressionType
+}
+
+// This will be false outside of the tests, so when building the server binary,
+// any code where you see `if msgTraceRunInTests` statement will be compiled
+// out, so this will have no performance penalty.
+var (
+	msgTraceRunInTests   bool
+	msgTraceCheckSupport bool
+)
+
+// Returns the message trace object, if message is being traced,
+// and `true` if message will be delivered to applications, or
+// `false` if only a trace events message needs to be produced.
+func (c *client) isMsgTraceEnabled() (*msgTrace, bool) {
```

Comment:
This should be inlined by compiler, can we double check?

Reply:
It is. See the lack of "CALL" and instead the code from msgtrace.go embedded in the deliverMsg() function:
```
TEXT github.com/nats-io/nats-server/v2/server.(*client).deliverMsg(SB) /Users/ik/dev/go/src/github.com/kozlovic/nats-server/server/client.go
  client.go:3336        0x1003596a0             f9400b90                MOVD 16(R28), R16                                                                       
  client.go:3336        0x1003596a4             d10803f1                SUB $512, RSP, R17                                                                      
  client.go:3336        0x1003596a8             eb10023f                CMP R16, R17                                                                            
  client.go:3336        0x1003596ac             5400a289                BLS 1300(PC)                                                                            
  client.go:3336        0x1003596b0             d10a03f4                SUB $640, RSP, R20                                                                      
  client.go:3336        0x1003596b4             a93ffa9d                STP (R29, R30), -8(R20)                                                                 
  client.go:3336        0x1003596b8             9100029f                MOVD R20, RSP                                                                           
  client.go:3336        0x1003596bc             d10023fd                SUB $8, RSP, R29                                                                        
  client.go:3336        0x1003596c0             f9015be4                MOVD R4, 688(RSP)                                                                       
  client.go:3336        0x1003596c4             f90167e7                MOVD R7, 712(RSP)                                                                       
  client.go:3336        0x1003596c8             f90173ea                MOVD R10, 736(RSP)                                                                      
  client.go:3336        0x1003596cc             f9017fed                MOVD R13, 760(RSP)                                                                      
  client.go:3336        0x1003596d0             39018bff                MOVB ZR, 98(RSP)                                                                        
  msgtrace.go:268       0x1003596d4             f9410410                MOVD 520(R0), R16                                                                       
  msgtrace.go:269       0x1003596d8             b4000070                CBZ R16, 3(PC)                                                                          
  msgtrace.go:272       0x1003596dc             3941a211                MOVBU 104(R16), R17                                                                     
  client.go:3338        0x1003596e0             14000003                JMP 3(PC)                                                                               
  client.go:3338        0x1003596e4             aa1f03f0                MOVD ZR, R16                  
..
```